include/revng/Model/Types.h

```diff
@@ -0,0 +1,903 @@
+  }
+
+  Type(llvm::StringRef NameRef, TypeKind::Values TK) :
+    UUID(0ULL), Name(NameRef), Kind(TK) {
+    auto Err = llvm::getRandomBytes(&UUID, sizeof(UUID));
+    if (Err)
+      UUID = 0ULL;
+  }
+
+  Type(llvm::StringRef NameRef) : Type(NameRef, TypeKind::Values::Invalid) {}
+  Type() : Type("") {}
+  ~Type() = default;
+
+  Type(const Type &) = default;
+  Type &operator=(const Type &) = default;
+
+  Type(Type &&) = default;
+  Type &operator=(Type &&) = default;
+
+  template<FlatCType T>
+  bool operator==(const T &Other) const {
+    return UUID == Other.UUID;
+  }
+
+public:
+  bool verify() const debug_function;
+
+private:
+  bool verifyBase() const debug_function;
+
+  std::strong_ordering compareBase(const Type &Other) const {
+    std::strong_ordering KindCmp = Kind <=> Other.Kind;
+    if (KindCmp != 0)
+      return KindCmp;
+
+    int NameCmp = llvm::StringRef(Name).compare(llvm::StringRef(Other.Name));
+    if (NameCmp < 0)
+      return std::strong_ordering::less;
+    if (NameCmp > 0)
+      return std::strong_ordering::greater;
+    return std::strong_ordering::equal;
+  }
+};
+// Enable introspection for flatc::Type
+INTROSPECTION_NS(flatc, Type, UUID, Name, Kind);
+
+// Make UpcastablePointers to flatc types usable in KeyedObject containers
+template<FlatCType T>
+struct KeyedObjectTraits<UpcastablePointer<T>> {
+
+  static std::string key(const UpcastablePointer<T> &Val) { return Val->Name; }
+
+  static UpcastablePointer<T> fromKey(const std::string &Name) {
+    return UpcastablePointer<T>(new T(Name));
+  }
+};
+
+/// \brief A qualifier for a flatc::Type
+struct flatc::Qualifier {
+  QualifierKind::Values Kind = QualifierKind::Invalid;
+  std::size_t Size = 0;
+
+  // Kind is not Invalid, Pointer and Const have no Size, Array has Size.
+  bool verify() const debug_function {
+    switch (Kind) {
+    case QualifierKind::Invalid:
+      return false;
+    case QualifierKind::Pointer:
+    case QualifierKind::Const:
+      return not Size;
+    case QualifierKind::Array:
+      return Size;
+    }
+    return false;
+  }
+
+  static Qualifier createConst() { return { QualifierKind::Const, 0 }; }
+
+  static Qualifier createPtr() { return { QualifierKind::Pointer, 0 }; }
+
+  static Qualifier createArray(std::size_t S) {
+    return { QualifierKind::Array, S };
+  }
+
+  bool operator==(const Qualifier &) const = default;
+};
+INTROSPECTION_NS(flatc, Qualifier, Kind, Size);
+
+namespace flatc {
+
+inline bool isConstQualifier(const Qualifier &Q) {
+  revng_assert(Q.verify());
+  return Q.Kind == QualifierKind::Const;
+}
+
+inline bool isArrayQualifier(const Qualifier &Q) {
+  revng_assert(Q.verify());
+  return Q.Kind == QualifierKind::Const;
+}
+
+inline bool isPtrQualifier(const Qualifier &Q) {
+  revng_assert(Q.verify());
+  return Q.Kind == QualifierKind::Const;
+}
+
+using TypesVector = SortedVector<UpcastablePointer<flatc::Type>>;
+
+} // end namespace flatc
+
+struct flatc::TypesTree {
```

Comment:
It is a tree?

```suggestion
struct flatc::TypesList {
```

Couldn't this just be field in `model::Binary`?

Reply:
It will become a field in `model::Binary`. I left it separated because during development I was testing it in isolation. This will soon go away.